storage/src/persistent/database.rs

```diff
@@ -82,61 +155,72 @@ pub trait KeyValueStoreWithSchema<S: KeyValueSchema> {
 
-    /// Insert new key value pair into WriteBatch.
+    /// Removes every element that predicate(elem) evaluates to false
     ///
     /// # Arguments
-    /// * `key` - Value of key specified by schema
-    /// * `value` - Value to be inserted associated with given key, specified by schema
-    fn put_batch(
-        &self,
-        batch: &mut WriteBatch,
-        key: &S::Key,
-        value: &S::Value,
-    ) -> Result<(), DBError>;
+    /// * `predicate` - functor used for assessment
+    fn retain(&self, predicate: &dyn Fn(&S::Key) -> bool) -> Result<(), DBError>;
 
     /// Write batch into DB atomically
     ///
     /// # Arguments
     /// * `batch` - WriteBatch containing all batched writes to be written to DB
-    fn write_batch(&self, batch: WriteBatch) -> Result<(), DBError>;
+    fn write_batch(&self, batch: Vec<(S::Key, S::Value)>) -> Result<(), DBError>;
+
+    /// Return memory usage statistics
+    ///
+    fn total_get_mem_usage(&self) -> Result<usize, DBError>;
 }
 
-pub trait GetInMemStats {
-    fn get_stats(&self) -> Result<RocksDBStats, DBError>;
+pub trait KeyValueStoreWithSchema<S: KeyValueSchema>:
+    KeyValueStoreBackend<S> + KeyValueStoreWithSchemaIterator<S>
+{
 }
 
-impl GetInMemStats for DB {
-    fn get_stats(&self) -> Result<RocksDBStats, DBError> {
-        let memory_usage_stats = rocksdb::perf::get_memory_usage_stats(Some(&[&self]), None)?;
-        Ok(RocksDBStats {
-            mem_table_total: memory_usage_stats.mem_table_total,
-            mem_table_unflushed: memory_usage_stats.mem_table_unflushed,
-            mem_table_readers_total: memory_usage_stats.mem_table_readers_total,
-            cache_total: memory_usage_stats.cache_total,
-        })
+impl<S: KeyValueSchema> KeyValueStoreWithSchemaIterator<S> for DB {
+    fn iterator(&self, mode: IteratorMode<S>) -> Result<IteratorWithSchema<S>, DBError> {
+        let cf = self
+            .cf_handle(S::name())
+            .ok_or(DBError::MissingColumnFamily { name: S::name() })?;
+
+        let iter = match mode {
+            IteratorMode::Start => self.iterator_cf(cf, rocksdb::IteratorMode::Start),
+            IteratorMode::End => self.iterator_cf(cf, rocksdb::IteratorMode::End),
+            IteratorMode::From(key, direction) => self.iterator_cf(
+                cf,
+                rocksdb::IteratorMode::From(&key.encode()?, direction.into()),
+            ),
+        };
+
+        Ok(IteratorWithSchema(iter, PhantomData))
+    }
+
+    fn prefix_iterator(&self, key: &S::Key) -> Result<IteratorWithSchema<S>, DBError> {
+        let key = key.encode()?;
+        let cf = self
+            .cf_handle(S::name())
+            .ok_or(DBError::MissingColumnFamily { name: S::name() })?;
+
+        Ok(IteratorWithSchema(
+            self.prefix_iterator_cf(cf, key),
+            PhantomData,
+        ))
     }
 }
 
-impl<S: KeyValueSchema> KeyValueStoreWithSchema<S> for DB {
+impl<S: KeyValueSchema> KeyValueStoreWithSchema<S> for DB {}
```

Comment:
@mat13mn 
do we need this empty line?
`impl<S: KeyValueSchema> KeyValueStoreWithSchema<S> for DB {}`